pkg/planner/core/plan_cache.go

```diff
@@ -59,16 +59,26 @@ func SetParameterValuesIntoSCtx(sctx base.PlanContext, isNonPrep bool, markers [
 	vars := sctx.GetSessionVars()
 	vars.PlanCacheParams.Reset()
 	for i, usingParam := range params {
-		val, err := usingParam.Eval(sctx.GetExprCtx().GetEvalCtx(), chunk.Row{})
-		if err != nil {
-			return err
-		}
-		if isGetVarBinaryLiteral(sctx, usingParam) {
-			binVal, convErr := val.ToBytes()
-			if convErr != nil {
-				return convErr
+		var (
+			val types.Datum
+			err error
+		)
+		if c, ok := usingParam.(*expression.Constant); ok {
+			// If the variable is readonly, it's folded to a constant value.
+			val = c.Value
+			intest.Assert(c.DeferredExpr == nil, "folded readonly user variable should not contain deferred expression")
```

Comment:
Can we add this Assert into the If condition like "if c, ok := usingParam.(*Constant); ok && c.DeferredExpr==nil"